### DISCUSSION TITLE
How to implement a seclist of seclist/list

### DISCUSSION THREAD

**Discussion by 1191000814** [lschoe/mpyc#82] [Q&A] [answered]
I want to create **a seclist** `L` which contains elements for seclist/list, because I want to access L with secure index, but it may be impossible because it make error when I get item of the seclist with sec_int index. So how can I do in other ways?

**Comment by lschoe** [comment, marked as answer]
Indeed, the elements of a `seclist` are currently required to be a secure number (secure type `mpc.SecFld()`, `mpc.SecInt()`, or `mpc.SecFxp()`). 

If you want a nested secure list, say a 2D list, where the inner lists are all of the same length `n`, you may try to convert 2D indices `(i,j)` to `n*i+j` to flatten the secure list of lists.

**Reply by 1191000814** [reply]
I use the 2d array for **a adjacency matrix of graph**. But sometimes I want to compute the degree of a node, that is  to calculate the sum of a row(**the row num is a secure int**) of a 2d matrix, how can I get it?

**Reply by lschoe** [reply]
Well, I would first compute the row sums of the adjacency matrix, and then obliviously select the row sum that you want. Something like
```python
 d = seclist([sum(A[n*i:n*(i+1)]) for i in range(n)])[r]
```
where `r` is a secure integer holding the secret row number, and A is the (flat) seclist representing the graph.

**Comment by 1191000814** [comment]
Hello, What does the attribute l (bit_length) of SecInt represent? Why can it still represent a large number when I set l=1? In addition, can setting a smaller l value reduce decoding time? Because I want to make my program computing with a large dataset faster.

**Reply by lschoe** [reply]
For `mpc.Secint(l)`, the parameter `l` represent the bit length for secure (signed) integers. This allows for secure integer arithmetic with correct results provided the secret-shared integer values remain between $-2^{l-1}$ and $2^{l -1}$. If the secret-shared values get outside this range, some programs will still work because of the so-called _headroom_ used with secure integers. The headroom is needed for secure comparisons and other operations: these operations will potentially give wrong results if the values are outside the above range.

By using smaller bit lengths `l` the performance of your program will improve. You can test to see at which `l` things start to fail. And then try to analyze why your program should still work around `l+1`, say.

**Comment by 1191000814** [comment]
Hi, if I use `ls = sec_int.array(np.array([1, 2, 3, 4]))` to create a secure array, How can I get a value with a secure index like `ls[sec_int(1)]`(I know this causes a error), So How to get it ?

**Reply by lschoe** [reply]
Secure indexing is only supported for seclists, not for secure Numpy arrays. Using `np_tolist()`and `np_fromlist()` you can convert between secure Numpy arrays and list of secure integers, and then also convert to seclists.

**Comment by 1191000814** [comment]
Hi again! If I want to use mpyc on **multiple machines**, rather than multiple terminals on one machine, how can I do it?

**Reply by lschoe** [reply]
Hi, for that check the documentation, e.g., at https://mpyc.readthedocs.io/en/latest/demos.html and search for "-P" to find examples on how to do this.
